### PR TITLE
[WIP] Authorization section

### PR DESCRIPTION
**Pull request by oberstet**

### CODE REVIEW

File: rfc/text/advanced/ap_52_authorization.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+At any moment, there can be zero, one or many *Session*s with different `sessionid|int`s authenticated under the _same_ *Principal*.
+
+*Session*s MAY be required to *Authorize* to perform a specific *Action* on an URI or URI pattern in a *Realm*.
+
+This distinction between *Authentication* and *Authorization* follows the established practice called "AAA":
+
+- **A**uthentication: Establishes who it is (**"subject"**)?
+- **A**uthorization: Decide in a *Realm* whether *Action* (**"operation"**) on URI or URI pattern (**"object"**) is allowed for the requesting *Principal* (**"subject"**)?
+- **A**uditability: Record for a *Realm* what *Action* (**"operation"**) on what URI or URI pattern (**"object"**) was requested by which *Principal* (**"subject"**), and was it allowed or denied?
```
Comment (KSDaemon) [inline rfc/text/advanced/ap_52_authorization.md:36]: Hmm never heard this version of a last A. 
What I know is:
AAA = Authentication, Authorization, Accounting. 
Looking at google's search results for `AAA` — almost all links are about `Accounting`. 

Yeah, it is the same thing. But I think it is a more broadly used version. So mb rename `Auditability` → `Accounting`?

**Review by KSDaemon — COMMENTED** [review]
Good! 👍🏻 

@ecorm can you also put your eyes on this?

File: rfc/text/advanced/ap_52_authorization.md (hunk 1 above)
Comment (oberstet) [reply rfc/text/advanced/ap_52_authorization.md:36]: I'll rename it. It seems to be more common with "accountability", even though it is wrong:

https://csrc.nist.gov/glossary/term/accountability
https://csrc.nist.gov/glossary/term/authentication_authorization_and_accounting

The third A is usually simply an "login log" (who sudo'ed what command and such).

I have never seen it implemented as "accountability" or "non-repudiation" - which exist, but not (usually) as part of AAA. Anyways, I'll change!

File: rfc/text/advanced/ap_52_authorization.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+* `authrole|string`
+* `authid|string`
+
+_for_ that *Session* running _in_ that *Client*.
+
+The triple `(realm, authrole, authid)` is called *Principal*, and the *Session* is authenticated under that *Principal*.
+
+At any moment, there can be zero, one or many *Session*s with different `sessionid|int`s authenticated under the _same_ *Principal*.
+
+*Session*s MAY be required to *Authorize* to perform a specific *Action* on an URI or URI pattern in a *Realm*.
```
[the remark this answers is not in the document]
Comment (ecorm) [reply rfc/text/advanced/ap_52_authorization.md:30]: This where the term "endpoint" applies, btw. E.g.
> ...perform a specific *Action* on an endpoint specified by a URI...

Just a remark, I don't want to start a big debate over this. :grin:

File: rfc/text/advanced/ap_52_authorization.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+This distinction between *Authentication* and *Authorization* follows the established practice called "AAA":
+
+- **A**uthentication: Establishes who it is (**"subject"**)?
```
Comment (ecorm) [inline rfc/text/advanced/ap_52_authorization.md:34]:
```suggestion
- **A**uthentication: Establishes who it is (**"subject"**)
```

File: rfc/text/advanced/ap_52_authorization.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+- **A**uthorization: Decide in a *Realm* whether *Action* (**"operation"**) on URI or URI pattern (**"object"**) is allowed for the requesting *Principal* (**"subject"**)?
```
Comment (ecorm) [inline rfc/text/advanced/ap_52_authorization.md:35]:
```suggestion
- **A**uthorization: Decides within a *Realm* whether an *Action* (**"operation"**) on an URI or URI pattern (**"object"**) is allowed for the requesting *Principal* (**"subject"**)
```

File: rfc/text/advanced/ap_52_authorization.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+- **A**ccounting: Record for a *Realm* what *Action* (**"operation"**) on what URI or URI pattern (**"object"**) was requested by which *Principal* (**"subject"**), and was it allowed or denied?
```
Comment (ecorm) [inline rfc/text/advanced/ap_52_authorization.md:36]:
```suggestion
- **A**ccounting: Records, for a *Realm*, what *Action* (**"operation"**) on what URI or URI pattern (**"object"**) was requested by which *Principal* (**"subject"**), and whether it was allowed or denied
```

**Review by ecorm — CHANGES_REQUESTED** [review]
This is good foundational stuff. Well done. My suggestions are all of a language style nature.

File: rfc/text/advanced/ap_52_authorization.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (oberstet) [reply rfc/text/advanced/ap_52_authorization.md:30]: You might be surprised, I highly value those discussions, even when they might seem nit picking!

Just a remark about my mental model:

* "perform action on URI or URI pattern" =>
* if action == call | publish => the URI must be fully qualified, and is matched by the broker against all current subscriptions, and for each matching subscription, the broker dispatches the published payload as an EVENT to the callees or subscribers on that registration or subscription, and each callee | subscriber is an XXX within a session with handling code attached in the app
* if action == register | subscribe => the URI can be fully qualified or pattern based. if successful, the respective registration | subscription is extended with the session

File: rfc/text/advanced/ap_52_authorization.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+*Authentication* lets the *Router* verify the identity of a *Session* often as a prerequisite to allowing access to resources in a *Realm*.
+
+When the *Session* authenticates to a *Router* successfully, the *Router* will have established
+
+* `realm|string`
+* `authrole|string`
+* `authid|string`
+
+_for_ that *Session* running _in_ that *Client*.
```
[the remark this answers is not in the document]
Comment (oberstet) [reply rfc/text/advanced/ap_52_authorization.md:24]: the principal is determined for the session which is code that runs in the client

**Comment by oberstet** [comment]
@ecorm fantastic! thank you so much=) I've merged all of you changes but 1. and I'm not too tied to my version, pls just pop up again if you want and I'll merge yours

**Comment by ecorm** [comment]
@oberstet The text looks fine the way it is now, so that suggested change you declined couldn't have been that important.

**Comment by oberstet** [comment]
alright, hooray=) let's merge (Konstantin also ack'ed already) this PR, so we have sth to build on (if required) in other parts or extend later

**Comment by SunFulong** [comment]
Sorry for missing this round, it seems more complete than previous discuss now!

**Pull request merged**